src/ode23s.jl

```diff
@@ -78,76 +80,165 @@ function start{O<:ExplicitODE,S<:ModifiedRosenbrockStepper}(s::Solver{O,S})
+    # unpack
     stepper = s.stepper
     ode     = s.ode
     step    = state.step
-    opts    = s.options
-
+    opts    = s.stepper.options
     F1, F2, J = state.F1, state.F2, state.J
-
+    k1,k2,k3,ynew =  state.k1, state.k2, state.k3, state.ynew
     t, dt, y, dy = step.t, state.dt, step.y, step.dy
-    # F!, J! = ode.F!, ode.J!
-    d, e32 = stepper.d, stepper.e32
-
+    F! = ode.F!
     F0 = dy
 
-    while true
+    # see whether we're done
+    if tdir(s)*t >= tdir(s)*opts.tstop
+        # nothing left to integrate
+        return finish
+    end
 
-        state.iters += 1
-        if state.iters > s.options.maxiters
-            return ((step.t,step.y), state)
-        end
+    # increment iteration counter
+    state.iters += 1
+    if state.iters > opts.maxiters
+        println("Reached maximum number of iterations $(opts.maxiters)")
+        return abort
+    end
 
-        # trim the step size to match the bounds of integration
-        dt = min(s.options.tstop-t,dt)
+    W = lufact!( eye(J) - dt*const_d*J )
 
-        W = lufact!( eye(J) - dt*d*J )
+    # Approximate time-derivative of F, we are using F1 as a
+    # temporary array
+    F!(t+dt/100,y,F1)
+    tder = 100*const_d*(F1-F0)
 
-        # Approximate time-derivative of F, we are using F1 as a
-        # temporary array
-        ode.F!(t+dt/100,y,F1)
-        tder = 100*d*(F1-F0)
+    # modified Rosenbrock formula
+    # TODO: update k1,k2,k3 in-place
+    k1[:] = W \ (F0 + tder)
+    F!(t+dt/2, y+dt*k1/2, F1)
+    k2[:] = W \ (F1 - k1) + k1
+    for i=1:length(y)
+        ynew[i] = y[i] + dt*k2[i]
+    end
+    F!(t+dt,   ynew,      F2)
+    k3[:] = W \ (F2 - const_e*(k2 - F1) - 2*(k1 - F0) + tder )
 
-        # modified Rosenbrock formula
-        # TODO: allocate some temporary space for these variables
-        k1 = W \ (F0 + tder)
-        ode.F!(t+dt/2, y+dt*k1/2, F1)
-        k2 = W \ (F1 - k1) + k1
-        ynew = y + dt*k2
-        ode.F!(t+dt,   ynew,      F2)
-        k3 = W \ (F2 - e32*(k2 - F1) - 2*(k1 - F0) + tder )
+    return cont
+end
 
-        delta = max(opts.reltol*max(opts.norm(y)::eltype(y),
-                                    opts.norm(ynew)::eltype(y)),
-                    opts.abstol) # allowable error
+function errorcontrol!{O<:ExplicitODE,S<:ModifiedRosenbrockStepper}(s::Solver{O,S}, state::RosenbrockState)
 
-        err = (dt/6)*(opts.norm(k1 - 2*k2 + k3)::eltype(y))/delta # error estimate
+    stepper = s.stepper
+    ode     = s.ode
+    step    = state.step
+    opts    = s.stepper.options
+    k1,k2,k3 =  state.k1, state.k2, state.k3
+    k1,k2,k3,ynew =  state.k1, state.k2, state.k3, state.ynew
+    t, dt, y, dy = step.t, state.dt, step.y, step.dy
 
-        # upon a failed step decrease the step size
-        dtnew = min(opts.maxstep,
-                    dt*0.8*err^(-1/3) )
+     # allowable error
+    delta = max(opts.reltol*max(opts.norm(y), opts.norm(ynew)),opts.abstol)
```

Comment:
I think we should optimize for 0.5, both the code and our time.  By the time this PR is merged 0.5 will be out.